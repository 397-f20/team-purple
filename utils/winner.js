const winner = (pollData) => {
  
  //creates an object that maps options to their scores along with a total value
  let sum = {};
  let lead = '';
  let max = 0;
  let data = Object.assign({}, pollData);
  let votes = Object.values(data.scores)
  //for each choices of each user's vote (based on db scheme)
  for (let vote of votes) {
    for (const [choice, scores] of Object.entries(vote)) {
      //first time accessing option? set it to that vote and create a total value
      if (!sum[choice]) {
        sum[choice] = Object.assign({}, scores);
        const tally = Object.values(scores);
        sum[choice].Total = tally.reduce((a, b) => {
          return a + b;
        }, 0);

        //check if we have a new winner
        if ((sum[choice].Total) > max) {
          max = (sum[choice].Total);
          lead = choice;
        }
        

      } 
      //else aggregate
      else {
        for (let criteria of data.criteria) {
          sum[choice][criteria] += scores[criteria]
          sum[choice].Total += scores[criteria]
        }

        //check if we have a new winner
        if ((sum[choice].Total) > max){
          max = (sum[choice].Total);
          lead = choice;
        }
      }
    }
  };

  //push the winner to the top of the lisr
  let res = [getResults(lead, sum, data.criteria)]

  //push everything else
  for (let option of data.options) {
    if (option != lead) {
      res.push(getResults(option, sum, data.criteria))
    }
  }
  return res;

};



const getResults = (option, sum, criteria) => {
  let scores = [];
  for (let crit of criteria) {
    scores.push(sum[option][crit]);
  };

  //might be a more efficient way to do this but this adds the total to the results graph
  scores.push(sum[option].Total)


  return {
    title: option,
    barData: {
      labels: [...criteria, 'Total'],
      datasets: [{
        data: scores,
      }, ],
    }
  };
}



export default winner;