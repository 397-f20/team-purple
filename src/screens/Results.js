import React, { useEffect, useState } from "react";
import {
  StyleSheet,
  SafeAreaView,
  ScrollView,
  Text,
  View,
  SectionList,
  Button,
  ActivityIndicator,
  FlatList,
  TouchableOpacity,
  Clipboard,
} from "react-native";
import { fonts, colors } from "../styles/all_styles";
import { firebase } from "../../utils/firebase";
import Header from "../components/01_Atoms/Header";
import winner from "../../utils/winner";
import Stars from "react-native-stars";
import { MaterialCommunityIcons } from "@expo/vector-icons";
import { Icon } from "react-native-elements";
import ResultSection from "../components/02_Molecules/ResultSection";

const Results = ({ route, navigation }) => {
  const [pollData, setPollData] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  const [win, setWin] = useState(null);
  const [copiedCode, setCopiedCode] = useState(false);
  const [voteCount, setVoteCount] = useState(route.params.count); // how many votes are in
  var roomCode = route.params.roomCode;
  let pollId;

  if (!route.params) {
    pollId = "-MKq65RirJuEavylaDax";
  } else {
    pollId = route.params.pollId;
  }

  // for copying room code to clipboard
  const copyToClipboard = () => {
    console.log("called copy");
    Clipboard.setString(roomCode);
    setCopiedCode(true);
  };

  // change back button to a home button
  React.useLayoutEffect(() => {
    navigation.setOptions({
      headerLeft: () => (
        <TouchableOpacity
          onPress={() => navigation.navigate("Home")}
          style={{ marginLeft: "10px" }}
        >
          <MaterialCommunityIcons name="home" size={25} />
        </TouchableOpacity>
      ),
      headerRight: () => (
        <View
          style={{
            display: "flex",
            flexDirection: "row",
            marginRight: "10px",
          }}
        >
          <Text style={fonts.p}>Room Code: {roomCode}</Text>
          <TouchableOpacity onPress={() => copyToClipboard()}>
            <Icon name="clipboard" type="material-community" />
          </TouchableOpacity>
        </View>
      ),
    });
  }, [navigation]);

  // getting all vote data from pollId
  useEffect(() => {
    setIsLoading(true);
    const db = firebase.database().ref("polls/" + pollId);

    const handleData = (snap) => {
      if (snap.val()) setPollData(snap.val());
      setIsLoading(false);
    };

    db.on("value", handleData, (error) => console.log(error));

    return () => {
      db.off("value", handleData);
    };
  }, []);

  // calls util function for placement of options
  useEffect(() => {
    if (pollData != null) setWin(winner(pollData));
  }, [pollData]);

  // logic for the vote count
  useEffect(() => {
    const db = firebase
      .database()
      .ref("polls/" + pollId)
      .child("count");

    const handleData = (snap) => {
      if (snap.val()) {
        setVoteCount(snap.val());
      }
    };

    db.on("value", handleData, (error) => console.log(error));
    return () => {
      db.off("value", handleData);
    };
  }, []);

  return (
    <SafeAreaView style={styles.container}>
      {win == null ? (
        <ActivityIndicator />
      ) : (
        <View style={styles.contentContainer}>
          <View style={styles.voteCountContainer}>
            {copiedCode ? (
              <Text style={{ textAlign: "center" }}>Room Code Copied!</Text>
            ) : null}
          </View>
          <View style={styles.voteCountContainer}>
            <Text style={fonts.p}> Vote Count: </Text>
            <Text style={[fonts.p, styles.voteCount]}> {voteCount} </Text>
          </View>
          <View style={styles.promptContainer}>
            <Text style={[fonts.h2]}>Prompt</Text>
            <Text style={fonts.p}>{pollData.prompt}</Text>
          </View>

          <Text style={[fonts.h2, { marginBottom: 10 }]}>Winner</Text>
          {win.map((option) => (
            (option.win)?
            <ResultSection data={option}/> : null))}

          <View style={styles.divider} />

          <Text style={[fonts.h2, { marginBottom: 10, color: "grey" }]}>
            Other Results
          </Text>
          {win.map((option) => (
            (!option.win)?
            <ResultSection headerColor={"#A9A9A9"} headerTextColor={"white"} data={option} other={true}/> : null
          ))}
        </View>
      )}
    </SafeAreaView>
  );
};

const buttonBase = {
  flex: 1,
  justifyContent: "center",
  alignItems: "center",
  height: 50,
  padding: 10,
};

const styles = StyleSheet.create({
  divider: {
    marginVertical: 20,
    marginBottom: 35,
    borderBottomColor: "grey",
    borderBottomWidth: 0.5,
    opacity: 0.5,
    width: "90%",
    alignSelf: "center",
  },
  container: {
    marginHorizontal: "5%",
    marginTop: 20,
    backgroundColor: "white",
    borderRadius: 10,
  },
  promptContainer: {
    marginBottom: 20,
  },
  contentContainer: {
    padding: 20,
  },
  roomCode: {
    ...buttonBase,
    backgroundColor: "#bababa",
    color: "white",
  },
  voteCountContainer: {
    width: "100%",
    display: "flex",
    flexDirection: "row",
    justifyContent: "flex-end",
  },
  voteCount: {
    color: colors.primaryColor,
    fontWeight: "bold",
  },
});

export default Results;
