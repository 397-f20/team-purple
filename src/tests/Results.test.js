import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import { NavigationContainer } from "@react-navigation/native";
import renderer from "react-test-renderer";
import winner from "../../utils/winner";
import Results from "../screens/Results";

// class router{
//   constructor(id) {
//     this.params = { pollId: id}
//   }
// }

// route = new router('fox');

// const route = () => {
//   return {
//     params: () => {
//       return {
//         pollId: () => {
//           return "fox";
//         },
//       };
//     },
//   };
// };

// const route = () => {
//   return {
//     params: {
//         pollId: "fox",
//     }
//   };
// };

// const route = () => {
//   this.params = () => {
//     return {pollId: "fox"}
//   }
// }

// describe("<Results />", () => {
//   it("has 1 child", async () => {
//     const tree = renderer.create(<Results route={route} />).toJSON();
//     await act(async () => {
//       expect(tree.children.length).toBe(1);
//     });
//   });
// });

const route = {
  params: { pollId: "-MKq65RirJuEavylaDax", roomCode: "piza", count: 21 },
};

const navigation = { navigate: jest.fn(), setOptions: jest.fn() };



import mockedWinner from "../../utils/winner";

const mockedinner = jest.fn(() =>  [{
      title: 'option',
      overall: 15,
      labels: ['hi','byee'],
      criteriaRatings: [2, 4],
      win: true
    }]
 );

 jest.mock("../../utils/winner", () => {
  
    return [{
      title: 'option',
      overall: 15,
      labels: ['hi','byee'],
      criteriaRatings: [2,4],
      win: true
    }];
});



describe("Hello", () => {
  it("renders the correct message", async () => {
    const { queryByText } =  render(
      <Results navigation={navigation} route={route} />
    );
    await expect(queryByText("Pie")).not.toBeNull();
  });
});

describe("<Results />", () => {
  it("has 1 child", async () => {
    const tree = renderer
      .create(<Results navigation={navigation} route={route} />)
      .toJSON();
    await expect(tree.length).toBe(1);
  });
});
