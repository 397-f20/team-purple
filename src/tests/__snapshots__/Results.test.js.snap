// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent should render correctly in "debug" mode 1`] = `
<ForwardRef(SafeAreaView)
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 10,
      "marginHorizontal": "5%",
      "marginTop": 20,
    }
  }
>
  <View>
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="small"
    />
    <Text>
      Pie
    </Text>
  </View>
</ForwardRef(SafeAreaView)>
`;
